yammbs/_store.py

```diff
@@ -542,24 +542,60 @@ def from_qcarchive_dataset(
 
         store = cls(database_name)
 
-        for qm_molecule in tqdm(dataset.qm_molecules, desc="Storing molecules"):
-            molecule = Molecule.from_mapped_smiles(qm_molecule.mapped_smiles, allow_undefined_stereo=True)
-            molecule.add_conformer(Quantity(qm_molecule.coordinates, "angstrom"))
+        # adapted from MoleculeRecord.from_molecule, MoleculeStore.store, and
+        # DBSessionManager.store_molecule_record
+        with store._get_session() as db:
+            # instead of DBSessionManager._smiles_already_exists
+            seen = set(db.db.query(DBMoleculeRecord.mapped_smiles))
+            for qm_molecule in tqdm(dataset.qm_molecules, desc="Storing molecules"):
+                if qm_molecule.mapped_smiles in seen:
+                    continue
+                seen.add(qm_molecule.mapped_smiles)
+                molecule = Molecule.from_mapped_smiles(qm_molecule.mapped_smiles, allow_undefined_stereo=True)
+                db_record = DBMoleculeRecord(
+                    mapped_smiles=qm_molecule.mapped_smiles,
+                    inchi_key=molecule.to_inchi(fixed_hydrogens=True),
+                )
+                db.db.add(db_record)
+                db.db.commit()
 
-            molecule_record = MoleculeRecord.from_molecule(molecule)
-            store.store(molecule_record)
+        # close the session here and re-open to make sure all of the molecule
+        # IDs have been flushed to the db
 
-            store.store_qcarchive(
-                QMConformerRecord(
-                    molecule_id=store.get_molecule_id_by_smiles(
-                        molecule_record.mapped_smiles,
-                    ),
-                    qcarchive_id=qm_molecule.qcarchive_id,
-                    mapped_smiles=qm_molecule.mapped_smiles,
-                    coordinates=qm_molecule.coordinates,
-                    energy=qm_molecule.final_energy,
+        # adapted from MoleculeStore.store_qcarchive,
+        # QMConformerRecord.from_qcarchive_record, and
+        # DBSessionManager.store_qm_conformer_record
+        with store._get_session() as db:
+            seen = set(
+                db.db.query(
+                    DBQMConformerRecord.qcarchive_id,
                 ),
             )
+            # reversed so the first record encountered wins out. this matches
+            # the behavior of the version that queries the db each time
+            smiles_to_id = {
+                smi: id
+                for id, smi in reversed(
+                    db.db.query(
+                        DBMoleculeRecord.id,
+                        DBMoleculeRecord.mapped_smiles,
+                    ).all(),
+                )
+            }
+            for record in tqdm(dataset.qm_molecules, desc="Storing Records"):
+                if record.qcarchive_id in seen:
+                    continue
```

Comment:
One of these lines isn't hit in code coverage - I think `record.qcarchive_id` is always unique, so skipping over duplicates isn't necessary (here), whereas this is not the case for molecules de-duplicated by SMILES?

Reply:
That sounds right, I'll just delete the check. Thanks!

I'm not sure how this works in the ORM, but possibly a `UNIQUE` constraint could be added on that field to the database itself if you wanted.

Reply:
I hadn't thought that far - do you think it makes sense to add it?

Reply:
I'm not sure. Knowing `UNIQUE` exists is near the edge of my database expertise. I think it will cause an error if multiple records with the same ID try to be inserted, so if other parts of the code rely on the uniqueness assumption, it might be a good thing to add. Otherwise I don't think it should have any effect.